src/boot/efi/cpio.c

```diff
@@ -0,0 +1,450 @@
+
+        assert(cpio_buffer);
+        assert(cpio_buffer_size);
+        assert_cc(sizeof(trailer) % 4 == 0);
+
+        a = ReallocatePool(*cpio_buffer, *cpio_buffer_size, *cpio_buffer_size + sizeof(trailer));
+        if (!a)
+                return EFI_OUT_OF_RESOURCES;
+
+        *cpio_buffer = a;
+        CopyMem((UINT8*) *cpio_buffer + *cpio_buffer_size, trailer, sizeof(trailer));
+        *cpio_buffer_size += sizeof(trailer);
+
+        return EFI_SUCCESS;
+}
+
+EFI_STATUS pack_cpio(
+                EFI_LOADED_IMAGE *loaded_image,
+                const CHAR16 *match_suffix,
+                const CHAR8 *target_dir_prefix,
+                UINT32 dir_mode,
+                UINT32 access_mode,
+                UINTN tpm_pcr,
+                const CHAR16 *tpm_description,
+                VOID **ret_buffer,
+                UINTN *ret_buffer_size) {
+
+        _cleanup_(FileHandleClosep) EFI_FILE_HANDLE root = NULL, extra_dir = NULL;
+        UINTN dirent_size = 0, buffer_size = 0, n_items = 0, n_allocated = 0;
+        _cleanup_freepool_ CHAR16 *loaded_image_path = NULL, *j = NULL;
+        _cleanup_freepool_ EFI_FILE_INFO *dirent = NULL;
+        _cleanup_(strv_freep) CHAR16 **items = NULL;
+        _cleanup_freepool_ VOID *buffer = NULL;
+        UINT32 inode = 1; /* inode counter, so that each item gets a new inode */
+        EFI_STATUS err;
+
+        assert(loaded_image);
+        assert(ret_buffer);
+        assert(ret_buffer_size);
+
+        root = LibOpenRoot(loaded_image->DeviceHandle);
+        if (!root)
+                return log_error_status_stall(EFI_LOAD_ERROR, L"Unable to open root directory.");
+
+        loaded_image_path = DevicePathToStr(loaded_image->FilePath);
+        if (!loaded_image_path)
+                return log_oom();
+
+        j = PoolPrint(L"%s" EXTRA_DIR_SUFFIX, loaded_image_path);
+        if (!j)
+                return log_oom();
+
+        err = open_directory(root, j, &extra_dir);
+        if (err == EFI_NOT_FOUND) {
+                /* No extra subdir, that's totally OK */
+                *ret_buffer = NULL;
+                *ret_buffer_size = 0;
+                return EFI_SUCCESS;
+        }
+        if (EFI_ERROR(err))
+                return log_error_status_stall(err, L"Failed to open extra directory of loaded image: %r", err);
+
+        for (;;) {
+                _cleanup_freepool_ CHAR16 *d = NULL;
+
+                err = readdir_harder(extra_dir, &dirent, &dirent_size);
+                if (EFI_ERROR(err))
+                        return log_error_status_stall(err, L"Failed to read extra directory of loaded image: %r", err);
+                if (!dirent) /* End of directory */
+                        break;
+
+                if (dirent->FileName[0] == '.')
+                        continue;
+                if (dirent->Attribute & EFI_FILE_DIRECTORY)
+                        continue;
+                if (match_suffix && !endswith_no_case(dirent->FileName, match_suffix))
```

Comment:
I feel like the first 3 checks here should be part of readdir_harder

Reply:
not sure that'd really make things ncier, since the invocations of readdir_harder() we have filter by different stuff...

dunno, maybe for a later cleanup patch, but i don't see it for now.